source/docs/h2_metadata.md

```diff
@@ -0,0 +1,111 @@
+### Overview
+
+Envoy provides a way for users to communicate extra information associating with a stream that is
+not carried in the standard HTTP(s) and HTTP/2 headers and payloads. The
+information is represented by a string key-value pair. For example, users can
+pass along RTT information associated with a stream using a key of "rtt info", and a value of
+"100ms". In Envoy, we call this type of information metadata.
+A stream can be associated with multiple metadatas, and the multiple metadatas
+are represented by a map.
+
+Note: the metadata implementation is still in progress, and the doc is in draft
+version.
+
+### Limitation
+
+For the ease of implementation and compatibility purpose, metadata will only be
+supported in HTTP/2. Metadata sent in any other protocol should result in protocol
+errors or be ignored.
+
+Envoy can consume or add metadata. But for one request (or response), we expect at most one of the
+two operations to be performed.
+TODO(soya3129): check if this limitation is needed after the implementation is
+complete.
+
+### Envoy metadata handling
+
+Envoy provides the functionality to proxy, process and add metadatas.
+
+## Proxy metadata
+
+(To be implemented)
+
+If not specified, all the metadata received by Envoy is proxied to the next hop
+unmodified. Note that, we do not guarantee the same frame order will be preserved from
+hop by hop. That is, metadata from upstream at the beginning of a stream can be
+received by the downstream at the end of the stream.
```

Comment:
So I agree that we don't guarantee anything with respect to body order (and doccing that up is an excellent idea!) but I think we can and probably should make sure we don't mess with end stream.  For example if we forward end-stream ahead of metadata in a response headed downstream that stream may be torn down before the metadata is received.  I think we have to make sure that if we've read metadata we defer end stream until metadata is flushed through the filter chain.  What do you think?